src/mongocxx/insert_many_builder.hpp

```diff
@@ -0,0 +1,65 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <mongocxx/config/prelude.hpp>
+
+#include <bsoncxx/stdx/optional.hpp>
+
+#include <mongocxx/bulk_write.hpp>
+#include <mongocxx/options/insert.hpp>
+#include <mongocxx/result/insert_many.hpp>
+#include <mongocxx/stdx.hpp>
+
+namespace mongocxx {
+MONGOCXX_INLINE_NAMESPACE_BEGIN
+
+///
+/// Class to build an insert_many bulk write operation.
+///
+/// This is used by mongocxx::collection for mongocxx::collection::insert_many.
+///
+class MONGOCXX_API insert_many_builder {
+   public:
+    ///
+    /// Initializes a new builder for an insert_many bulk write operation.
+    ///
+    /// @param options
+    ///   The options for this insert_many.
+    ///
+    explicit insert_many_builder(const options::insert& options);
+
+    ///
+    /// Appends a single insert to insert_many bulk write operation. The document is copied into the
+    /// bulk operation completely, so there is no dependency between the life of the document and
+    /// the insert_many bulk write operation itself.
```

Comment:
When append is called on the bulk write, it creates a new scoped_bson_t. I am pretty sure that doing that copies all of the contents. It might also be copied again inside libmongoc. We should look into that.